include/forward_index_builder.hpp

```diff
@@ -0,0 +1,288 @@
+        return os;
+    }
+
+    static std::ostream &write_header(std::ostream &os, uint32_t document_count)
+    {
+        return write_document(os, &document_count, std::next(&document_count));
+    }
+
+    [[nodiscard]] static auto batch_file(std::string const &output_file,
+                                         std::ptrdiff_t     batch_number) noexcept
+        -> std::string
+    {
+        std::ostringstream os;
+        os << output_file << ".batch." << batch_number;
+        return os.str();
+    }
+
+    struct Batch_Process {
+        std::ptrdiff_t               batch_number;
+        std::vector<Record>          records;
+        Document_Id                  first_document;
+        std::string const &          output_file;
+    };
+
+    void run(Batch_Process bp, process_term_function_type process_term) const
+    {
+        auto basename = batch_file(bp.output_file, bp.batch_number);
+
+        std::ofstream os(basename);
+        std::ofstream title_os(basename + ".documents");
+        std::ofstream url_os(basename + ".urls");
+        std::ofstream term_os(basename + ".terms");
+        write_header(os, bp.records.size());
+
+        std::map<std::string, uint32_t> map;
+
+        for (auto const &record : bp.records) {
+            title_os << record.trecid() << '\n';
+
+            auto content = record.content();
+            std::vector<uint32_t> term_ids;
+
+            auto process = [&](auto &term) {
+                term = process_term(std::move(term));
+                uint32_t id = 0;
+                if (auto pos = map.find(term); pos != map.end()) {
+                    id = pos->second;
+                } else {
+                    id = map.size();
+                    map[term] = id;
+                    term_os << term << '\n';
+                }
+                term_ids.push_back(id);
+            };
+            std::istringstream content_stream(content);
+            std::string term;
+            while (content_stream >> term) {
+                process(term);
+            }
+            write_document(os, term_ids.begin(), term_ids.end());
+        }
+        logger() << "[Batch " << bp.batch_number << "] Processed documents [" << bp.first_document
+                 << ", " << (bp.first_document + bp.records.size()) << ")\n";
+    }
+
+    void merge(std::string const &basename,
+               std::ptrdiff_t     document_count,
+               std::ptrdiff_t     batch_count) const
+    {
+        using term_map_type =
+            std::map<std::string, std::unordered_map<std::ptrdiff_t, std::ptrdiff_t>>;
+        std::ofstream title_os(basename + ".documents");
+        std::ofstream url_os(basename + ".urls");
+        std::ofstream term_os(basename + ".terms");
+
+        logger() << "Merging titles\n";
+        for (auto batch : enumerate(batch_count)) {
+            std::ifstream title_is(batch_file(basename, batch) + ".documents");
+            title_os << title_is.rdbuf();
+        }
```

Comment:
Does it mean we need to be able to load all the `.documents` batches in memory? 
Fine for now, but we should think about a k-way merge solution in the next release. `Stxxl` (which is already a dependency) will make all this very easy!

Reply:
No, you load only one at a time, and redirect to another file. Note that `title_is` has a local scope, it will be closed after each iteration.

Reply:
Same goes for URLs. What we **have to store in memory** are terms at this point. Because we have an in-memory map. That could be maybe improved at a later time.